src/basic/virt.c

```diff
@@ -485,6 +485,50 @@ int detect_virtualization(void) {
         return r;
 }
 
+int detect_private_users(void) {
+        static thread_local int cached = -1;
+
+        if (cached < 0) {
+                _cleanup_fclose_ FILE *f = NULL;
+                _cleanup_free_ char *buf = NULL;
+                size_t n_allocated = 0;
+                ssize_t n;
+                uint32_t a, b, c;
+                int r;
+
+                f = fopen("/proc/self/uid_map", "re");
+                if (!f)
+                        return log_debug_errno(errno, "Failed to open /proc/self/uid_map: %m");
+
+                n = getline(&buf, &n_allocated, f);
+                if (n < 0)
+                        return log_debug_errno(errno, "Failed to read /proc/self/uid_map: %m");
+
+                r = sscanf(buf, "%"PRIu32" %"PRIu32" %"PRIu32, &a, &b, &c);
+                if (r < 3)
+                        return log_debug_errno(errno, "Failed to parse /proc/self/uid_map: %m");
+
+                if (a > 0 || b > 0 || c != UINT32_MAX) {
+                        cached = true;
+                        goto exit;
+                }
+
+                n = getline(&buf, &n_allocated, f);
+                if (feof(f))
+                        cached = false;
+                else if (n < 0)
+                        return log_debug_errno(errno, "Failed to read /proc/self/uid_map: %m");
+                else
+                        /* we have multiple lines of mapping */
+                        cached = true;
+        }
+ exit:
+        if (cached)
+                return detect_virtualization();
```

Comment:
I really wouldn't do this fallback checking of virtualization. I think userns and virtualization are orthogonal: you can have userns used by containers, but userns can also be used for other purposes that i wouldn't want to use the term "containers" or "virtualization" for, for example systemd's own services where PrivateUsers=yes is set...

I'd much rather make this an option that works like `systemd-detect-virt --chroot`, i.e. that is entirely orthogonal to container/virtualization detection, and really only checks whether userns is used or not. 

(Note that we used to consider "chroot" a kind of container env, but I dropped that quite some time ago, since I really wouldn't want to call it container env, and I think "userns" is any different here)